src/streamlink/plugins/onetv.py

```diff
@@ -31,72 +23,32 @@ def can_handle_url(cls, url):
-
-        return update_scheme(self.url, url)
-
     def hls_session(self):
-        res = self.session.http.get(update_scheme(self.url, self._session_api))
+        res = self.session.http.get("http://stream.1tv.ru/get_hls_session")
         data = self.session.http.json(res)
         # the values are already quoted, we don't want them quoted
         return {k: unquote(v) for k, v in data.items()}
 
-    @property
-    def is_live(self):
-        m = self._url_re.match(self.url)
-        return m and m.group("live") is not None
-
-    def vod_data(self, vid=None):
-        """
-        Get the VOD data path and the default VOD ID
-        :return:
-        """
-        page = self.session.http.get(self.url)
-        m = self._vod_re.search(page.text)
-        vod_data_url = m and urljoin(self.url, m.group(0))
-        if vod_data_url:
-            log.debug("Found VOD data url: {0}".format(vod_data_url))
-            res = self.session.http.get(vod_data_url)
-            return self.session.http.json(res)
-
     def _get_streams(self):
-        if self.is_live:
-            log.debug("Loading live stream for {0}...".format(self.channel))
-
-            res = self.session.http.get(self.live_api_url, data={"r": random.randint(1, 100000)})
-            live_data = self.session.http.json(res)
-
-            # all the streams are equal for each type, so pick a random one
-            hls_streams = live_data.get("hls")
-            if hls_streams:
-                url = random.choice(hls_streams)
-                url = url + '&' + urlencode(self.hls_session())  # TODO: use update_qsd
-                yield from HLSStream.parse_variant_playlist(self.session, url, name_fmt="{pixels}_{bitrate}").items()
-
-            mpd_streams = live_data.get("mpd")
-            if mpd_streams:
-                url = random.choice(mpd_streams)
-                yield from DASHStream.parse_manifest(self.session, url).items()
-
-        elif self.channel == "1tv":
-            log.debug("Attempting to find VOD stream for {0}...".format(self.channel))
-            vod_data = self.vod_data()
-            if vod_data:
-                log.info(f"Found VOD: {vod_data[0]['title']}")
-                for stream in vod_data[0]['mbr']:
-                    yield stream['name'], HTTPStream(self.session, update_scheme(self.url, stream['src']))
+        url = self.session.http.get(
+            "https://stream.1tv.ru/api/playlist/1tvch_as_array.json",
+            data={"r": random.randint(1, 100000)},
+            schema=validate.Schema(
+                validate.transform(parse_json),
+                {"hls": [validate.url()]},
+                validate.get("hls"),
+                validate.get(0),
+            ))
+
+        if not url:
+            return
+
+        if "georestrictions" in url:
+            log.error("Stream is geo-restricted")
+            return
+
+        url = url + "&" + urlencode(self.hls_session())
```

Comment:
Use `update_qsd` instead of concatenating the URL string

Reply:
> 400 Client Error

`update_qsd` won't work here, `urlencode` will make the url invalid
https://github.com/streamlink/streamlink/blob/44d4afaa931e0e3f525ae9ec7adcb91f1020acb2/src/streamlink/utils/url.py#L103

Reply:
> `urlencode` will make the url invalid

How? Can you show an example URL?

`update_qsd` doesn't have test cases for proper URL encoding, but I don't see any issues with the current implementation.
Example tests for correctly encoded and for incorrecly encoded input URLs:

```diff
diff --git a/tests/test_utils_url.py b/tests/test_utils_url.py
index eb87e6df..2e0e14f4 100644
--- a/tests/test_utils_url.py
+++ b/tests/test_utils_url.py
@@ -43,3 +43,5 @@ def test_update_qsd():
     assert update_qsd("http://test.se?&two=", {"one": ''}, keep_blank_values=False) == "http://test.se?one=", \
         "should set one blank"
     assert update_qsd("http://test.se?one=", {"two": 2}) == "http://test.se?one=&two=2"
+    assert update_qsd("http://test.se?foo=%20&bar=%3F", {"foo": "!"}) == "http://test.se?foo=%21&bar=%3F"
+    assert update_qsd("http://test.se?foo= &bar=?", {"foo": "!"}) == "http://test.se?foo=%21&bar=%3F"
```

Reply:
> How? Can you show an example URL?

`update_qsd` makes `cdn` invalid and the server doesn't accept it.

```
--- valid
<Response [200]>

https://mobdrm.1tv.ru/hls-live10/streams/1tv/1tv.m3u8?cdn=https://edge1.1internet.tv&netprav=/hls-vod11/streams/1tv/netprav.m3u8&e=1621770289&s=%2FmI3ltgPPHFCpQ


--- invalid
<Response [400]>
ERROR: Not valid CDN

https://mobdrm.1tv.ru/hls-live10/streams/1tv/1tv.m3u8?cdn=https%3A%2F%2Fedge1.1internet.tv&netprav=%2Fhls-vod11%2Fstreams%2F1tv%2Fnetprav.m3u8&e=1621770289&s=%2FmI3ltgPPHFCpQ
```

Reply:
`urlencode` has an option `safe=''` with which it is possible that individual characters are not escaped:
```
urlencode({"cdn": "http://test.com"}, safe="/:")
'cdn=http://test.com'

```
`update_qsd` would have to be adapted accordingly.

Reply:
> `update_qsd` would have to be adapted accordingly.

#3746